fortitude/src/rules/correctness/error_handling.rs

```diff
@@ -0,0 +1,337 @@
+                    Ok(StatType::IoStat)
+                } else {
+                    Err(anyhow!("Unknown routine: {identifier_text}"))
+                }
+            }
+            _ => Err(anyhow!("Node does not have a stat type")),
+        }
+    }
+}
+
+enum CheckStatus {
+    Checked,
+    Unchecked,
+    Overwritten,
+}
+
+/// ## What does it do?
+/// This rule detects whether a `stat`, `iostat`, and `cmdstat` variable is checked
+/// within the same scope it is set.
+///
+/// ## Why is this bad?
+/// By default, `allocate` statements will crash the program if the allocation
+/// fails. This is often the desired behaviour, but to provide for cases in
+/// which the user wants to handle allocation errors gracefully, they may
+/// optionally check the status of an `allocate` statement by passing a variable
+/// to the `stat` argument:
+///
+/// ```f90
+/// allocate (x(BIG_INT), stat=status)
+/// if (status /= 0) then
+///   call handle_error(status)
+/// end if
+/// ```
+///
+/// However, if the `stat` variable is not checked, the program will continue to
+/// run despite the allocation failure, which can lead to undefined behaviour.
+/// Similar behaviour is exhibited by `deallocate` and IO statements such as
+/// `open`, `read`, and `close`.
+///
+/// To avoid confusing and bug-prone control flow, the checks on status parameters
+/// should occur within the same scope in which they are set.
+#[derive(ViolationMetadata)]
+pub(crate) struct UncheckedStat {
+    name: String,
+    stat: StatType,
+    result: CheckStatus,
+}
+
+impl Violation for UncheckedStat {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let Self { name, stat, result } = self;
+        let check_status = match result {
+            CheckStatus::Checked => "checked",
+            CheckStatus::Unchecked => "not checked",
+            CheckStatus::Overwritten => "overwritten before being checked",
+        };
+        format!("{stat} argument '{name}' is {check_status} in this scope.")
+    }
+}
+
+impl AstRule for UncheckedStat {
+    fn check(_settings: &Settings, node: &Node, source: &SourceFile) -> Option<Vec<Diagnostic>> {
+        let src = source.source_text();
+
+        // Check this is an error checking statement, and get the stat type
+        let stat_type = StatType::from_node(node, src).ok()?;
+        let stat_name: &'static str = stat_type.into();
+
+        // Find a 'stat' argument in the allocate statement
+        let arg_list = if matches!(node.kind(), "subroutine_call" | "call_expression") {
+            node.child_with_name("argument_list")?
+        } else {
+            *node
+        };
+        let stat_node = arg_list.named_children(&mut node.walk()).find(|child| {
+            child.kind() == "keyword_argument"
+                && child.child_by_field_name("name").is_some_and(|n| {
+                    n.to_text(src).map(|s| s.to_lowercase()) == Some(stat_name.to_string())
+                })
```

Comment:
Do we already have a helper function for keyword args? Might be nice to add one if not!

Reply:
We have `is_keyword_argument` in `ast.rs`, but I think we can generalise it a little and make it part of the `FortitudeNode` trait. I'll see what I can do.

Reply:
Whoops, just pushed a fix to https://github.com/PlasmaFAIR/fortitude/pull/458/commits/476d5c23fa90b6c4d77fc8838e22f00005b100b4! I went with a trio of functions:
- `node.kwarg()` -> the `keyword_argument` node
- `node.kwarg_value()` -> the `value` node of the kwarg
- `node.kwarg_exists()` -> utility that just calls `.is_some()` on the result of `kwarg`
I cleaned up a few other bits of the code with them too.

Reply:
Nice, looks great!